File: others/cpp/199.Binary Tree Right Side View.cpp
/**
 * Source: LeetCode  199. Binary Tree Right Side View
 * Author: ceezyyy
 * Date: 10-09-2019
 */


 /*
 Medium

 Given a binary tree, imagine yourself standing on the right side of it, 
 return the values of the nodes you can see ordered from top to bottom.

 Example:

 Input: [1,2,3,null,5,null,4]
 Output: [1, 3, 4]

 Explanation:

	1            <---
  /   \
 2     3         <---
  \     \
   5     4       <---

 */


#include<vector>
using namespace std;

 /**
  * Definition for a binary tree node.
  * struct TreeNode {
  *     int val;
  *     TreeNode *left;
  *     TreeNode *right;
  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  * };
  */
class Solution {
public:
	struct TreeNode {
		int val;
		TreeNode *left;
		TreeNode *right;
		TreeNode(int x) : val(x), left(NULL), right(NULL) {}
	};

	vector<int> rightSideView(TreeNode* root) {
		vector<int> res;
		queue<TreeNode*> q;
		if (!root)
			return res;
		q.push(root);  // the first element
		while (!q.empty()) {
			vector<int> temp_vec;  // the similar with level order traversal
			int temp_size = q.size();  // the size of each level
			for (int i = 0; i < temp_size; i++) {
				TreeNode* node = q.front();
				q.pop();
				temp_vec.push_back(node->val);
				if (node->left)  // left child
					q.push(node->left);
				if (node->right)  // right child
					q.push(node->right);
			}
			res.push_back(temp_vec.back());  // the last element is the right side node
		}
		return res;
	}
};


/**
 * Runtime: 4 ms, faster than 72.00% of C++ online submissions for Binary Tree Right Side View.
 * Memory Usage: 9.7 MB, less than 75.68% of C++ online submissions for Binary Tree Right Side View.
 */

File: others/Python3/199.Binary Tree Right Side View.py
"""
@Title: 199. Binary Tree Right Side View
@Tag: binary tree
@Date: Mar-06 2020
@Author: ceezyyy
@Difficulty: Medium
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        queue, res = [root], []
        while queue:
            n = len(queue)
            for i in range(n):
                node = queue.pop(0)
                if i == n - 1:
                    res.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return res


"""
Runtime: 28 ms, faster than 80.95% of Python3 online submissions for Binary Tree Right Side View.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Right Side View.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
